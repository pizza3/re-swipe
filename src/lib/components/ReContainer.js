import React, { Component } from "react";
import { CheckMark, Cross } from "../assets/icon";
class ReContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      arr: [],
      displayChildren: [],
      maxElement: this.props.max,
    };
  }

  componentDidMount() {
    this.renderChildren();
  }

  renderChildren = () => {
    const { children } = this.props;
    let arr = this.createChildren(children);
    let displayChildren = [];
    // only adds maxElement children only
    displayChildren = arr.slice(
      children.length - this.state.maxElement,
      children.length
    );
    this.setState({
      arr: arr,
      children: children,
      displayChildren
    });
  };

  // create each card component and pass down neccessary props
  createChildren = children => {
    return React.Children.toArray(children).map((child, i) => {
      return React.cloneElement(child, {
        key: i,
        num: i,
        mass: this.props.mass,
        damping: this.props.damping,
        handleDown: this.handleDown,
        handleUp: this.handleUp,
        updateChildren: this.updateChildren,
        handleOnSwipe: this.handleOnSwipe
      });
    });
  };

  handleOnSwipe = (swipeDirection, metaData) => {
    const { onSwipe } = this.props;
    if (onSwipe) {
      onSwipe(swipeDirection, metaData);
    }
  };

  updateChildren = () => {
    let { arr, maxElement, displayChildren } = this.state;
    maxElement += 1;
    displayChildren.pop();
    if (this.props.children.length >= maxElement) {
      displayChildren.unshift(arr[this.props.children.length - maxElement]);
      this.setState({
        displayChildren,
        maxElement
      });
    } else {
      this.setState({
        displayChildren,
        maxElement
      });
    }
  };

  render() {
    const style = {
      position: "relative",
      width: "100%",
      height: "100%",
      overflow: "hidden"
    };
    const iconContainer = {
      position: "absolute",
      left: "0px",
      right: "0px",
      marginLeft: "auto",
      marginRight: "auto",
      width: "350px",
      bottom: "50px"
    };
    const icon = {
      width: 0,
      margin: 0,
      padding: 0,
      height: 0
    };
    return (
      <>
        <div style={style}>
          <>{this.state.displayChildren}</>
        </div>
        <span style={iconContainer}>
          <div style={icon}>{Cross()}</div>
        <div style={{...icon,float:'right'}}>{CheckMark()}</div>
        </span>
      </>
    );
  }
}

ReContainer.defaultProps = {
  mass: 0.7,
  damping: 0.8,
  trigger: false,
  max: 3,
  onSwipe: undefined
};

export default ReContainer;
