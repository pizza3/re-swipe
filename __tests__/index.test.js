import React from "react";
import { configure, shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
import ReContainer, { ReCard } from "../src/lib/index";

configure({
  adapter: new Adapter()
});

const App = () => {
  const foo = [1, 2, 3, 4, 5];
  const parentStyle = {
    position: "relative",
    width: "100%",
    height: "100vh"
  };
  // NOTE: it is important to enclose <ReContainer/> component inside a parent div, being an absolute container it will inherit the parent dimensions (width & height).
  return (
    <div style={parentStyle}>
      <ReContainer>
        {foo.map((value, index) => {
          return <ReCard key={index}>{value}</ReCard>;
        })}
      </ReContainer>
    </div>
  );
};

const wrapper = shallow(<App/>);

describe("<ReContainer/>", () => {
  it("Creates basic snapshot", () => {
    wrapper;
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
