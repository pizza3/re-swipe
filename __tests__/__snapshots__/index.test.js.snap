// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReContainer/> Creates basic snapshot 1`] = `
<div
  style={
    Object {
      "height": "100vh",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <ReContainer
    damping={0.8}
    mass={0.7}
    max={5}
    trigger={false}
  >
    <ReCard
      height={400}
      key="0"
      metaData={Object {}}
      width={300}
    >
      1
    </ReCard>
    <ReCard
      height={400}
      key="1"
      metaData={Object {}}
      width={300}
    >
      2
    </ReCard>
    <ReCard
      height={400}
      key="2"
      metaData={Object {}}
      width={300}
    >
      3
    </ReCard>
    <ReCard
      height={400}
      key="3"
      metaData={Object {}}
      width={300}
    >
      4
    </ReCard>
    <ReCard
      height={400}
      key="4"
      metaData={Object {}}
      width={300}
    >
      5
    </ReCard>
  </ReContainer>
</div>
`;
